### ISSUE TITLE
Wrong EOS Masking ?

### ISSUE THREAD

**Issue by sarvam31** [epfl-dlab/transformers-CFG#103]
I noticed that eos_token_id is unmasked only when no other token_id is unmasked (referring to `get_next_token_acceptance_for_single_stack` in `IncrementalTokenRecognizer`)

But suppose we have a grammar to generate 1 or 2 digit number expressed as
`root ::= [0-9] ( [0-9] | )`

Now top of stack allows generating 0-9 which is expected but after 1 time step when a token for single digit is sampled by llm stack updates to point to next 0-9 only and no stack entry accounts for epsilon. So, we cannot generate 1 digit number even if grammar allows it.

I thought it should be handled in a way where another stack entry is present, corresponding to epsilon, which points to next possible byte that can be generated if epsilon is chosen for current rule and in case epsilon is one of the last alternates in grammar then unmask eos token too.

Please let me know if I understood it right or you need further information.

**Comment by Saibo-creator** [comment]
@sarvam31 , thank you for raising this issue and yes your observation is correct and your suggestion of fix is also correct. 

I fixed this in the recent PR #108 , 

Now the output of your grammar is like this


```terminal
Step 0
'1' (29508): 1.3009755611419678
'0' (29502): 0.7234944105148315
'5' (29550): -0.03773176670074463
'8' (29551): -0.14857470989227295
'2' (29518): -0.16474050283432007
'4' (29549): -0.2031155228614807
'3' (29538): -0.22310882806777954
'6' (29552): -0.4381660223007202
'9' (29542): -0.5209637880325317
'7' (29555): -0.6026785373687744
'<s>' (1): -inf
'<unk>' (0): -inf

Step 1
'0' (29502): 21.700206756591797
'1' (29508): 19.28874397277832
'9' (29542): 5.83770751953125
'7' (29555): 5.586031913757324
'6' (29552): 5.486726760864258
'5' (29550): 4.907416343688965
'8' (29551): 4.843409061431885
'3' (29538): 4.205896854400635
'4' (29549): 3.0457682609558105
'2' (29518): 1.9729324579238892
'</s>' (2): -12.689655303955078
'<unk>' (0): -inf

Step 2
'</s>' (2): -9.426843643188477
'[control_9]' (11): -inf
'[/INST]' (4): -inf
'[/AVAILABLE_TOOLS]' (7): -inf
'<s>' (1): -inf
'<unk>' (0): -inf
'[INST]' (3): -inf
'[AVAILABLE_TOOLS]' (6): -inf
'[TOOL_RESULTS]' (8): -inf
'[control_8]' (10): -inf
'[/TOOL_RESULTS]' (9): -inf
'[TOOL_CALLS]' (5): -inf
```

Here is the script to reproduce


```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from transformers_cfg.grammar_utils import IncrementalGrammarConstraint
from transformers_cfg.generation.logits_process import GrammarConstrainedLogitsProcessor

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model_id = "mistralai/Mistral-7B-Instruct-v0.3"

tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.pad_token = tokenizer.eos_token
model = AutoModelForCausalLM.from_pretrained(model_id, device_map="auto")
model.eval()

grammar_str = 'root ::= [0-9] ( [0-9] | )'
grammar_constraint = IncrementalGrammarConstraint(grammar_str, "root", tokenizer)
logits_processor = GrammarConstrainedLogitsProcessor(grammar_constraint, execution_mode="full_mask")

prompt = "Generate an arbitrary binary string."
messages = [{"role": "user", "content": prompt}]
input_ids = tokenizer.apply_chat_template(messages, add_generation_prompt=True, return_tensors="pt").to(model.device)

output = model.generate(
    input_ids,
    max_new_tokens=4,
    eos_token_id=tokenizer.eos_token_id,
    pad_token_id=tokenizer.pad_token_id,
    do_sample=True,
    logits_processor=[logits_processor],
    return_dict_in_generate=True,
    output_scores=True,
)

output_ids = output.sequences[:, input_ids.shape[-1] :]
output_str = tokenizer.decode(output_ids[0], skip_special_tokens=True)
print(f"Output: {output_str}")
print()

generation_index = 0 

top_k = 12

output_scores = torch.stack(output.scores).transpose(0, 1)
for i in range(output_ids.shape[1]):
    top_scores, top_indices = torch.topk(output_scores[generation_index][i], k=top_k)
    print(f"Step {i}")
    for score, id in zip(top_scores, top_indices):
        # if the id == the output_id, it means that the token was selected, print in green
        if id == output_ids[generation_index, i]:
            print(f"\033[92m{repr(tokenizer.decode(id))} ({id}): {score.item()}\033[0m")
        else:
            print(f"{repr(tokenizer.decode(id))} ({id}): {score.item()}")
    print()
```
